devito/mpi/halo_scheme.py

```diff
@@ -39,54 +40,53 @@ class HaloScheme(object):
+    ``loc_indices`` could be ``{t: t-1}``.
 
-    The dict ``loc_indices`` tells how to access/insert the halo along the
-    keyed Function's non-halo indices. For example, consider the Function
-    ``u(t, x, y)``. Assume ``x`` and ``y`` require a halo exchange. The
-    question is: once the halo exchange is performed, at what offset in ``t``
-    should it be placed? should it be at ``u(0, ...)`` or ``u(1, ...)`` or even
-    ``u(t-1, ...)``? ``loc_indices`` has as many entries as non-halo
-    dimensions, and each entry provides symbolic information about how to
-    access the corresponding non-halo dimension. Thus, in this example
-    ``loc_indices`` could be, for instance, ``{t: 0}`` or ``{t: t-1}``.
+    ``halos`` is a list of 2-tuples ``(Dimension, DataSide)``. This is metadata
+    about the halo exchanges, such as the Dimensions along which a halo exchange
+    is expected to be performed.
 
     Parameters
     ----------
-    exprs : tuple of IREq
-        The expressions for which the HaloScheme is built
+    exprs : expr-like or list of expr-like
+        The expressions for which the HaloScheme is derived.
     ispace : IterationSpace
-        Description of iteration directions and sub-iterators used in ``exprs``.
-    fmapper : dict, optional
-        The format is the same as ``M``. When provided, ``exprs`` and ``ispace``
-        are ignored. It should be used to aggregate several existing HaloSchemes
-        into a single, "bigger" HaloScheme, without performing any further analysis.
+        The iteration space of the expressions.
     """
 
-    def __init__(self, exprs=None, ispace=None, fmapper=None):
-        if fmapper is not None:
-            self._mapper = frozendict(fmapper.copy())
-            return
-
+    def __init__(self, exprs, ispace):
+        # Derive halo exchanges
         self._mapper = {}
         scope = Scope(exprs)
-        for f, v in hs_classify(scope).items():
-            halos = [Halo(*i) for i, hl in v.items() if hl is STENCIL]
+        for f, (halos, local) in classify(scope).items():
             if halos:
-                # There is some halo to be exchanged; *what* are the local
-                # (i.e., non-halo) indices?
-                dims = [i for i, hl in v.items() if hl is NONE]
-                loc_indices = hs_comp_locindices(f, dims, ispace, scope)
-
+                loc_indices = compute_local_indices(f, local, ispace, scope)
                 self._mapper[f] = HaloSchemeEntry(frozendict(loc_indices),
                                                   frozenset(halos))
-
-        # A HaloScheme is immutable, so let's make it hashable
         self._mapper = frozendict(self._mapper)
 
+        # Track the IterationSpace offsets induced by SubDomains/SubDimensions.
+        # These should be honored in the derivation of the `omapper`
+        self._honored = {}
+        for i in ispace.intervals:
+            if not i.dim._defines & set(self.dimensions):
```

Comment:
How come we need `set` here?

Reply:
well ``self.dimensions `` returns tuple, and intersection (``&``) is only defined over sets